libs/stats/tests/test_gm_ls_bitmask.py

```diff
@@ -31,10 +31,26 @@ def dataset():
         "y": np.linspace(0, 5, band_pq.shape[1]),
         "spec": index,
     }
-
-    data_vars = {"band_red": (("spec", "y", "x"), band_red), "QA_PIXEL": (("spec", "y", "x"), band_pq)}
-    attrs = dict(crs="epsg:32633", grid_mapping="spatial_ref")
-    xx = xr.Dataset(data_vars=data_vars, coords=coords, attrs=attrs)
+    pq_flags_definition = {'snow': {'bits': 5, 'values': {'0': 'not_high_confidence', '1': 'high_confidence'}},
```

Comment:
Are you sure you can't get this from the ODC?

Reply:
It should be accessible via something like `xx.pixel_quality.flags_definition`, yep

Reply:
> I just want to say that I find change to "dict" a bit confusing. For starters, `closing` and `opening` both use `dilation` and `erosion` internally (applied in different order), also unlike tuple, values in dictionary do not have native order, so it's harder to understand what the operation will do when looking at the call site of the function.
> 
> If you need to generalize this code to support longer chains of morphological operations, then I would suggest to use something like this
> 
> ```python
> def mask_cleanup(
>     mask: xr.DataArray,
>     filter: Iterable[Tuple[str, int]] = [("opening", 2), ("dilation", 5)],
>     name: Optional[str] = None
> ) -> xr.DataArray:
> ```
> 
> So you supply operations you want performed, in the order you want them performed. This allows to actually use any of this operations more than once. And reads much better at call site. Even better use enumeration instead of strings for `opening, closing, dilation, erosion`.

I like this suggestion if we can actually infer the order and type of operation from the parameter supplied. I agree that in the end it's just a series of erosion and dilation options. Closing of 5, opening of 5 and dilation of 5 would be the same as dilation of 5, erosion of 10 and dilation of 10.

Reply:
This is to build a new dataset object (e.g. xx) as an input dataset for geomad/pq testing. 

So I was thinking to get the definition from product unless there is a better option?

Reply:
@fangfy order is known, since `Iterable` is an ordered sequence, it could be a list or a tuple. So the question is what's the best way to represent "operataion+parameter", options would be

- `Tuple[str, int]`
- `Tuple[Enum, int]`
- Custom class

Reply:
> I just want to say that I find change to "dict" a bit confusing. For starters, `closing` and `opening` both use `dilation` and `erosion` internally (applied in different order), also unlike tuple, values in dictionary do not have native order, so it's harder to understand what the operation will do when looking at the call site of the function.
> 
> If you need to generalize this code to support longer chains of morphological operations, then I would suggest to use something like this
> 
> ```python
> def mask_cleanup(
>     mask: xr.DataArray,
>     filter: Iterable[Tuple[str, int]] = [("opening", 2), ("dilation", 5)],
>     name: Optional[str] = None
> ) -> xr.DataArray:
> ```
> 
> So you supply operations you want performed, in the order you want them performed. This allows to actually use any of this operations more than once. And reads much better at call site. 

> I like this suggestion if we can actually infer the order and type of operation from the parameter supplied. I agree that in the end it's just a series of erosion and dilation options. Closing of 5, opening of 5 and dilation of 5 would be the same as dilation of 5, erosion of 10 and dilation of 10.

@fangfy @Kirill888 I agree, the advantage of generalising it and allowing the user to define order is that it would allow more flexibility to go from something like `opening > dilation` as an initial test, to optimising to something like `erosion > dilation` for performance reasons once they're happy the first option does what they want

Reply:
> @fangfy order is known, since `Iterable` is an ordered sequence, it could be a list or a tuple. So the question is what's the best way to represent "operataion+parameter", options would be
> 
> * `Tuple[str, int]`
> * `Tuple[Enum, int]`
> * Custom class

true, I guess if you provide "binary_closing" instead of "closing", you can directly apply and chain the operations? 
(BTW did I leave comment in the wrong thread? sorry not sure how that happened... )

Reply:
All are very good suggestions. As this PR is not targeting this work, I have raise new issue ticket here - https://github.com/opendatacube/odc-tools/issues/300 so we can discuss it separately :)

Reply:
sounds good. I suggest we merge this PR and worry about improving that later.